internal/handlers/logs_handlers.go

```diff
@@ -7,115 +7,94 @@ import (
+	functionSig := chi.URLParam(r, "functionSig")
+	handleLogsRequest(w, r, contractAddress, functionSig)
 }
 
-func GetEventsWithContract(w http.ResponseWriter, r *http.Request) {
+func handleLogsRequest(w http.ResponseWriter, r *http.Request, contractAddress, functionSig string) {
 	chainId, err := api.GetChainId(r)
 	if err != nil {
 		api.BadRequestErrorHandler(w, err)
 		return
 	}
-	contractAddress := chi.URLParam(r, "contractAddress")
+
 	queryParams, err := api.ParseQueryParams(r)
 	if err != nil {
 		api.BadRequestErrorHandler(w, err)
 		return
 	}
 
-	var response = api.QueryResponse{
-		Meta: api.Meta{
-			ChainIdentifier: chainId,
-			ContractAddress: contractAddress,
-			FunctionSig:     "todo",
-			Page:            1,
-			Limit:           100,
-			TotalItems:      0,
-			TotalPages:      0,
-		},
-		Data: []interface{}{queryParams},
-	}
-
-	w.Header().Set("Content-Type", "application/json")
-	err = json.NewEncoder(w).Encode(response)
+	mainStorage, err := getMainStorage()
 	if err != nil {
-		log.Error().Err(err).Msg("Error encoding response")
+		log.Error().Err(err).Msg("Error creating storage connector")
 		api.InternalErrorHandler(w)
 		return
 	}
 
-}
-
-func GetEventsWithContractAndSignature(w http.ResponseWriter, r *http.Request) {
-	chainId, err := api.GetChainId(r)
+	logs, err := mainStorage.GetLogs(storage.QueryFilter{
+		FilterParams:    queryParams.FilterParams,
+		GroupBy:         queryParams.GroupBy,
+		SortBy:          queryParams.SortBy,
+		SortOrder:       queryParams.SortOrder,
+		Page:            queryParams.Page,
+		Limit:           queryParams.Limit,
+		Aggregates:      queryParams.Aggregates,
+		ContractAddress: contractAddress,
+		FunctionSig:     functionSig,
+	})
 	if err != nil {
-		api.BadRequestErrorHandler(w, err)
-		return
-	}
-	contractAddress := chi.URLParam(r, "contractAddress")
-	functionSig := chi.URLParam(r, "functionSig")
-	queryParams, err := api.ParseQueryParams(r)
-	if err != nil {
-		api.BadRequestErrorHandler(w, err)
+		log.Error().Err(err).Msg("Error querying logs")
+		api.InternalErrorHandler(w)
 		return
 	}
 
-	var response = api.QueryResponse{
+	response := api.QueryResponse{
 		Meta: api.Meta{
 			ChainIdentifier: chainId,
 			ContractAddress: contractAddress,
 			FunctionSig:     functionSig,
-			Page:            1,
-			Limit:           100,
-			TotalItems:      0,
-			TotalPages:      0,
+			Page:            queryParams.Page,
+			Limit:           queryParams.Limit,
+			TotalItems:      0, // TODO: Implement total items count
+			TotalPages:      0, // TODO: Implement total pages count
 		},
-		Data: []interface{}{queryParams},
+		Data: []interface{}{logs},
 		Aggregations: map[string]interface{}{
-			"count":         100,
-			"sum_value":     "1000000000000000000000",
-			"avg_gas_price": "20000000000",
+			"aggregates": queryParams.Aggregates,
 		},
 	}
 
+	sendJSONResponse(w, response)
+}
+
+func getMainStorage() (storage.IMainStorage, error) {
```

Comment:
Do you want to create a new client for every endpoint or should it be 1 for the API?

Reply:
done